lib/iris/tests/unit/data_manager/test_DataManager.py

```diff
@@ -0,0 +1,602 @@
+        self.assertIs(result, dm._lazy_array)
+
+
+class Test_data__getter(tests.IrisTest):
+    def setUp(self):
+        shape = (2, 3, 4)
+        size = np.cumprod(shape)[-1]
+        self.real_array = np.arange(size).reshape(shape)
+        self.lazy_array = da.from_array(self.real_array, chunks=shape)
+        self.mask_array = ma.masked_array(self.real_array)
+        self.mask_array_masked = self.mask_array.copy()
+        self.mask_array_masked[0, 0, 0] = ma.masked
+        self.realised_dtype = self.mask_array.dtype
+        self.lazy_mask_array = as_lazy_data(self.mask_array)
+        self.lazy_mask_array_masked = as_lazy_data(self.mask_array_masked)
+
+    def test_with_real_array(self):
+        dm = DataManager(self.real_array)
+        self.assertTrue(dm.has_real_data())
+        self.assertFalse(dm.has_lazy_data())
+        result = dm.data
+        self.assertTrue(dm.has_real_data())
+        self.assertFalse(dm.has_lazy_data())
+        self.assertIs(result, self.real_array)
+
+    def test_with_lazy_array(self):
+        dm = DataManager(self.lazy_array)
+        self.assertFalse(dm.has_real_data())
+        self.assertTrue(dm.has_lazy_data())
+        result = dm.data
+        self.assertTrue(dm.has_real_data())
+        self.assertFalse(dm.has_lazy_data())
+        self.assertArrayEqual(result, self.real_array)
+
+    def test_with_lazy_mask_array__not_masked(self):
+        dm = DataManager(self.lazy_mask_array,
+                         realised_dtype=self.realised_dtype)
+        self.assertFalse(dm.has_real_data())
+        self.assertTrue(dm.has_lazy_data())
+        result = dm.data
+        self.assertTrue(dm.has_real_data())
+        self.assertFalse(dm.has_lazy_data())
+        self.assertIsInstance(result, np.core.ndarray)
+        self.assertIsNone(dm._realised_dtype)
+        self.assertEqual(dm.dtype, self.realised_dtype)
+        self.assertArrayEqual(result, self.real_array)
+
+    def test_with_lazy_mask_array__masked(self):
+        dm = DataManager(self.lazy_mask_array_masked,
+                         realised_dtype=self.realised_dtype)
+        self.assertFalse(dm.has_real_data())
+        self.assertTrue(dm.has_lazy_data())
+        result = dm.data
+        self.assertTrue(dm.has_real_data())
+        self.assertFalse(dm.has_lazy_data())
+        self.assertIsInstance(result, ma.MaskedArray)
+        self.assertIsNone(dm._realised_dtype)
+        self.assertEqual(dm.dtype, self.realised_dtype)
+        self.assertArrayEqual(result, self.lazy_mask_array_masked)
+
+
+class Test_data__setter(tests.IrisTest):
+    def test_zero_ndim_real_with_scalar_int(self):
+        value = 456
+        dm = DataManager(np.array(123))
+        self.assertTrue(dm.has_real_data())
+        self.assertFalse(dm.has_lazy_data())
+        dm.data = value
+        self.assertTrue(dm.has_real_data())
+        self.assertFalse(dm.has_lazy_data())
+        self.assertArrayEqual(dm.data, np.array(value))
```

Comment:
Yeah I had made this comment before I read the rest of your tests. I thought it would be a good thing cause then you're being consistent regardless of whether it returns a lazy array or real array but that doesn't really fit what you're testing 
I'm happy with how you've done ti.